Source/WebCore/css/parser/CSSPropertyParser.cpp

```diff
@@ -5373,46 +5368,23 @@ bool CSSPropertyParser::consumeSystemFont(bool important)
     return true;
 }
 
+// FIXME: Share more code with consumeFontRaw.
 bool CSSPropertyParser::consumeFont(bool important)
 {
-    // Let's check if there is an inherit or initial somewhere in the shorthand.
-    CSSParserTokenRange range = m_range;
-    while (!range.atEnd()) {
-        CSSValueID id = range.consumeIncludingWhitespace().id();
-        if (id == CSSValueInherit || id == CSSValueInitial)
-            return false;
-    }
-
-    // Optional font-style, font-variant, font-stretch and font-weight.
+    // Optional font-style, font-variant, font-stretch and font-weight, in any order.
     RefPtr<CSSFontStyleValue> fontStyle;
     RefPtr<CSSPrimitiveValue> fontVariantCaps;
     RefPtr<CSSPrimitiveValue> fontWeight;
     RefPtr<CSSPrimitiveValue> fontStretch;
-
     while (!m_range.atEnd()) {
-        CSSValueID id = m_range.peek().id();
-        if (!fontStyle) {
-            fontStyle = consumeFontStyle(m_range, m_context.mode, CSSValuePool::singleton());
-            if (fontStyle)
-                continue;
-        }
-        if (!fontVariantCaps && (id == CSSValueNormal || id == CSSValueSmallCaps)) {
-            // Font variant in the shorthand is particular, it only accepts normal or small-caps.
-            // See https://drafts.csswg.org/css-fonts/#propdef-font
-            fontVariantCaps = consumeFontVariantCSS21(m_range);
-            if (fontVariantCaps)
-                continue;
-        }
-        if (!fontWeight) {
-            fontWeight = consumeFontWeight(m_range);
-            if (fontWeight)
-                continue;
-        }
-        if (!fontStretch) {
-            fontStretch = consumeFontStretchKeywordValue(m_range, CSSValuePool::singleton());
-            if (fontStretch)
-                continue;
-        }
+        if (!fontStyle && (fontStyle = consumeFontStyle(m_range, m_context.mode, CSSValuePool::singleton())))
+            continue;
+        if (!fontVariantCaps && (fontVariantCaps = consumeIdent<CSSValueNormal, CSSValueSmallCaps>(m_range)))
+            continue;
+        if (!fontWeight && (fontWeight = consumeFontWeight(m_range)))
+            continue;
+        if (!fontStretch && (fontStretch = consumeFontStretchKeywordValue(m_range, CSSValuePool::singleton())))
+            continue;
```

Comment:
Pre-existing bug, but this code actually breaks for cases like `normal normal normal 300 16px Arial` since it can't disambiguate `normal` if the non-normal values are not the order specified by the loop. Whereas `normal normal 300 normal 16px Arial` works. (This stood out to me because we had a similar bug with `list-style`, see https://bugs.webkit.org/show_bug.cgi?id=243991)

Testcase: `data:text/html,<span style="font: normal normal normal 300 16px Arial">Hello</span>`
Other browsers render this properly.

Reply:
I’m happy to fix that; I know exactly how, it is relatively easy. But should we use a different pull request for it? Or do you want me to add to this one?

Also, do you think other browsers will tolerate extra “normal” or does it count them?

Reply:
Separate PR would probably be better.